pkg/gator/bench/bench.go

```diff
@@ -0,0 +1,500 @@
+func runBenchmark(
+	engine Engine,
+	templates []*unstructured.Unstructured,
+	constraints []*unstructured.Unstructured,
+	reviewObjs []*unstructured.Unstructured,
+	opts *Opts,
+) (*Results, error) {
+	ctx := context.Background()
+	var setupBreakdown SetupBreakdown
+	var skippedTemplates []string
+	var skippedConstraints []string
+	loadedTemplateKinds := make(map[string]bool)
+
+	// Create the client for this engine
+	setupStart := time.Now()
+	clientStart := time.Now()
+	client, err := makeClient(engine, opts.GatherStats)
+	if err != nil {
+		return nil, fmt.Errorf("creating client: %w", err)
+	}
+	setupBreakdown.ClientCreation = time.Since(clientStart)
+
+	// Add templates (with skip support for incompatible templates)
+	templateStart := time.Now()
+	for _, obj := range templates {
+		templ, err := reader.ToTemplate(scheme, obj)
+		if err != nil {
+			return nil, fmt.Errorf("converting template %q: %w", obj.GetName(), err)
+		}
+		_, err = client.AddTemplate(ctx, templ)
+		if err != nil {
+			// Check if this is an engine compatibility issue
+			if isEngineIncompatibleError(err) {
+				skippedTemplates = append(skippedTemplates, obj.GetName())
+				continue
+			}
+			return nil, fmt.Errorf("adding template %q: %w", templ.GetName(), err)
+		}
+		// Track the constraint kind this template creates
+		loadedTemplateKinds[templ.Spec.CRD.Spec.Names.Kind] = true
+	}
+	setupBreakdown.TemplateCompilation = time.Since(templateStart)
+
+	// Check if all templates were skipped
+	loadedTemplateCount := len(templates) - len(skippedTemplates)
+	if loadedTemplateCount == 0 {
+		return nil, fmt.Errorf("no templates compatible with %s engine (all %d templates skipped)", engine, len(templates))
+	}
+
+	// Add constraints (skip those whose template was skipped)
+	constraintStart := time.Now()
+	for _, obj := range constraints {
+		kind := obj.GetKind()
+		if !loadedTemplateKinds[kind] {
+			skippedConstraints = append(skippedConstraints, obj.GetName())
+			continue
+		}
+		if _, err := client.AddConstraint(ctx, obj); err != nil {
+			return nil, fmt.Errorf("adding constraint %q: %w", obj.GetName(), err)
+		}
+	}
+	setupBreakdown.ConstraintLoading = time.Since(constraintStart)
+
+	// Check if all constraints were skipped
+	loadedConstraintCount := len(constraints) - len(skippedConstraints)
+	if loadedConstraintCount == 0 {
+		return nil, fmt.Errorf("no constraints loaded (all %d constraints skipped due to missing templates)", len(constraints))
+	}
+
+	// Add all objects as data (for referential constraints)
+	// Note: CEL driver doesn't support referential constraints, so we track skipped objects
+	dataStart := time.Now()
+	var skippedDataObjects []string
+	for _, obj := range reviewObjs {
+		_, err := client.AddData(ctx, obj)
+		if err != nil {
+			// CEL engine doesn't support referential data, so we track skipped objects
+			// and continue. The review will still work for non-referential constraints.
+			if engine == EngineCEL {
```

Comment:
why call addData for CEL engine and confirm error? why not just skip data loading if engine is CEL?

Reply:
It provides explicit tracking of which objects would have been used for referential constraints
Users get visibility into what's being skipped via SkippedDataObjects
The warning helps users understand that referential constraint benchmarks won't be accurate with CEL